tfx_addons/firebase_publisher/runner.py

```diff
@@ -0,0 +1,210 @@
+"""
+import glob
+import os
+import tempfile
+from typing import Any, Dict, List, Tuple
+
+import firebase_admin
+import tensorflow as tf
+from absl import logging
+from firebase_admin import credentials, ml
+from firebase_admin.ml import ListModelsPage, TFLiteModelSource
+from tfx.dsl.io import fileio
+from tfx.utils import io_utils
+
+_SIZE_LIMIT_MB = 80
+
+
+def prepare_fb_download_model(app_name: str, credential_path: str,
+                              storage_bucket: str, model_path: str,
+                              options: Dict[str, Any]) -> str:
+  """initialize Firebase app, and download the
+     target model to a temporary directory"""
+  tmp_dir = tempfile.gettempdir()
+  credential = None
+
+  if credential_path is not None:
+    tmp_credential_path = os.path.join(tmp_dir, "credentials.json")
+    io_utils.copy_file(credential_path, tmp_credential_path)
+    credential = credentials.Certificate(tmp_credential_path)
+    logging.info(
+        "credentials are copied into a temporary directory in local filesystem"
+    )
+
+  options["storageBucket"] = storage_bucket
+
+  firebase_admin.initialize_app(credential=credential,
+                                options=options,
+                                name=app_name)
+  logging.info("firebase app initialization is completed")
+
+  tmp_model_path = os.path.join(tmp_dir, "model")
+  io_utils.copy_dir(model_path, tmp_model_path)
+
+  return tmp_model_path
+
+
+def get_model_path_and_type(tmp_model_path) -> Tuple[bool, str]:
+  """get model path and flag if the model is TFLite"""
+  tflite_files = glob.glob(f"{tmp_model_path}/**/*.tflite")
+  is_tflite = len(tflite_files) > 0
+  model_path = tflite_files[0] if is_tflite else tmp_model_path
+
+  return is_tflite, model_path
+
+
+def upload_model(is_tflite: bool, model_path: str) -> TFLiteModelSource:
+  """upload model to GCS. If SavedModel, it will be
+     converted to TFLite under the hood"""
+  if is_tflite:
+    source = ml.TFLiteGCSModelSource.from_tflite_model_file(model_path)
+  else:
+    source = ml.TFLiteGCSModelSource.from_saved_model(model_path)
+
+  return source
+
+
+def check_model_size(source: TFLiteModelSource):
+  """the model size to be hosted in Firebase ML is limited
+      to 40mb max. if the size exceeds, RuntimeError is raised"""
+  gcs_path_for_uploaded_file = source.as_dict().get('gcsTfliteUri')
+  with tf.io.gfile.GFile(gcs_path_for_uploaded_file) as f:
+    file_size_in_mb = f.size() / (1 << 20)
+
+  if file_size_in_mb > _SIZE_LIMIT_MB:
+    fileio.remove(gcs_path_for_uploaded_file)
+    raise RuntimeError(
+        f"the file size {file_size_in_mb} exceeds"
+        f"the limit of {_SIZE_LIMIT_MB}. Uploaded file is removed.")
+
+
+def model_exist(model_list: ListModelsPage):
+  return len(model_list.models) > 0
+
+
+def update_model(model_list: ListModelsPage, source: TFLiteModelSource,
+                 tags: List[str], model_version: str):
+  """update existing model"""
```

Comment:
```suggestion
  """Updates existing model."""
```

The purpose of this utility is not clear to me.